libs/metagen/src/mdk_python/mod.rs

```diff
@@ -0,0 +1,338 @@
+
+use garde::external::compact_str::CompactStringExt;
+use heck::ToPascalCase;
+
+use crate::interlude::*;
+use crate::mdk::*;
+use crate::*;
+
+use self::utils::Memo;
+use self::utils::TypeGenerated;
+
+mod types;
+mod utils;
+
+#[derive(Serialize, Deserialize, Debug, garde::Validate)]
+pub struct MdkPythonGenConfig {
+    #[serde(flatten)]
+    #[garde(dive)]
+    pub base: crate::config::MdkGeneratorConfigBase,
+}
+
+impl MdkPythonGenConfig {
+    pub fn from_json(json: serde_json::Value, workspace_path: &Path) -> anyhow::Result<Self> {
+        let mut config: mdk_python::MdkPythonGenConfig = serde_json::from_value(json)?;
+        config.base.path = workspace_path.join(config.base.path);
+        config.base.typegraph_path = config
+            .base
+            .typegraph_path
+            .as_ref()
+            .map(|path| workspace_path.join(path));
+        Ok(config)
+    }
+}
+
+pub struct Generator {
+    config: MdkPythonGenConfig,
+}
+
+impl Generator {
+    pub const INPUT_TG: &'static str = "tg_name";
+    pub fn new(config: MdkPythonGenConfig) -> Result<Self, garde::Report> {
+        use garde::Validate;
+        config.validate(&())?;
+        Ok(Self { config })
+    }
+}
+
+impl crate::Plugin for Generator {
+    fn bill_of_inputs(&self) -> HashMap<String, GeneratorInputOrder> {
+        [(
+            Self::INPUT_TG.to_string(),
+            if let Some(tg_name) = &self.config.base.typegraph_name {
+                GeneratorInputOrder::TypegraphFromTypegate {
+                    name: tg_name.clone(),
+                }
+            } else if let Some(tg_path) = &self.config.base.typegraph_path {
+                GeneratorInputOrder::TypegraphFromPath {
+                    path: tg_path.clone(),
+                    name: self.config.base.typegraph_name.clone(),
+                }
+            } else {
+                unreachable!()
+            },
+        )]
+        .into_iter()
+        .collect()
+    }
+
+    fn generate(
+        &self,
+        inputs: HashMap<String, GeneratorInputResolved>,
+    ) -> anyhow::Result<GeneratorOutput> {
+        // return Ok(GeneratorOutput(Default::default()))
+        let tg = match inputs
+            .get(Self::INPUT_TG)
+            .context("missing generator input")?
+        {
+            GeneratorInputResolved::TypegraphFromTypegate { raw } => raw,
+            GeneratorInputResolved::TypegraphFromPath { raw } => raw,
+        };
+        let mut mergeable_output: IndexMap<PathBuf, Vec<RequiredObjects>> = IndexMap::new();
+
+        let mut tera = tera::Tera::default();
+        tera.add_raw_template("main_template", include_str!("static/main.py.jinja"))?;
+        tera.add_raw_template("types_template", include_str!("static/types.py.jinja"))?;
+        tera.add_raw_template("struct_template", include_str!("static/struct.py.jinja"))?;
+
+        let stubbed_funs = filter_stubbed_funcs(tg, &["python".to_string()])?;
+        for fun in &stubbed_funs {
```

Comment:
Not sure how this would work in python but what I meant bout the graphql queries is the variables and the data returned from queries. It'd be nice to be able to convert the JSONs into the concrete types to ease interacting with them. I don't mean auto conversion through the `gql` abstraction but manually by the user which I still think is very valuable.  This'll come in very handy when things change during regeneration, breaking the static analysis.